### ISSUE TITLE
Infinite loop in `lookupOrLoad` on errors.

### ISSUE THREAD

**Issue by gatorcse** [davenverse/mules#330]
I think we found a pretty critical bug in `lookupOrLoad`. When the lookup action results in an error, lookupOrLoad will attempt the action again. If it always results in an error, it will keep recursively attempting the action, resulting in an infinite loop.

Here's a snippet of the relevant code:
```scala
    stuffToGetTheCache
      .flatMap{
        case Some(s) => s.item.get.flatMap{
          case Left(_) => insertAtomic(k, action) >> lookupOrLoad(k, action) // 👈 Here's the issue.
          case Right(v) => F.pure(v)
        }
        case None => insertAtomic(k, action) >> lookupOrLoad(k, action)
      }
```

At the indicated line, it attempts to get the value in the cache. The value is `Either` the Throwable from the previous lookup attempt, or the found value. If there's an error in there (Left), _it fetches the value again_, and then _recursively calls itself_. If the action is always an error (ie, http4sclient.expect[Thing] is an error if it results in 404), it will keep calling itself, storing the error and calling itself again.

Not sure what the _desired_ behavior is here. I can see two likely expected use cases.
* On error, store the error like now, but next time we lookup then do attempt to get the value. This would mean that errors were effectively not cached, while still maintaining the `DispatchOne` functionality
* Cache the error values and re-propagating the error each time it's requested. This would mean that it would not attempt another lookup until the cache expires.

**Comment by ChristopherDavenport** [comment]
That seems a bit problematic. I am thinking that we should instead do something like this. I was working on it as the way this works isn't as great as what we can do here. Which is allow the delegates to all rely on the lookup if it fails with this mechanism.

https://github.com/davenverse/single-fibered/blob/main/core/src/main/scala/io/chrisdavenport/singlefibered/SingleFibered.scala

**Comment by gatorcse** [comment]
Woah, I was thinking about something along those lines.. "Everyone share this op" seems like a very useful abstraction. Would still need to integrate that into mules for this bug.

But outside of mules, is that lib "ready" for use? I have other ideas for that...

**Comment by gatorcse** [comment]
you got a branch around where you're messing around with that where I can help?

**Comment by henryxparker** [comment]
Can confirm this infinite loop just caused me a headache in some unit tests. Is there a reason we can't use the above PR? Or are choosing to wait for a PR that uses the `SingleFibered` style solution?

**Comment by ChristopherDavenport** [comment]
This is fixed, but as I worried ends up possibly causing it to call more than once in the error state. So this is a heads up. Will work to fix with the more robust solution for 0.5 series.
